tests/listpkg.rs

```diff
@@ -158,3 +140,91 @@ where
+        Some("basic") => {
+            list_dependencies_basic_child_main();
+        }
+        Some("workspace") => {
+            list_dependencies_workspace_child_main();
+        }
+        Some("alternate_registry_restriction_in_workspace") => {
+            list_dependencies_with_alternate_registry_restriction_in_workspace_child_main()
+        }
+        Some("alternate_registry_restriction_in_workspace_with_unset_alt") => {
+            list_dependencies_with_alternate_registry_restriction_in_workspace_unset_alt_child_main()
+        }
+        _ => {
+            parent_main();
+        }
+    }
+}
+
+fn parent_main() {
+    let test_cnt = 4_usize;
+    println!("running {} tests", test_cnt);
+
+    rust_child_test_with_assertion_fn(
+        "list_dependencies_basic_workspace",
+        TestVariants::Basic,
+        |output| {
+            let stderr = String::from_utf8_lossy(&output.stderr).to_string();
+            let lines: Vec<_> = stderr.lines().collect();
+
+            assert!(lines[0].starts_with("basic"));
+        },
+    );
+
+    rust_child_test_with_assertion_fn(
+        "list_dependencies_workspace",
+        TestVariants::Workspace,
+        |output| {
+            let stderr = String::from_utf8_lossy(&output.stderr).to_string();
+            let lines: Vec<_> = stderr.lines().collect();
+
+            if lines[0].starts_with("build1") {
+                assert!(lines[1].starts_with("dep1"));
+            } else {
+                assert!(lines[0].starts_with("dep1"));
+                assert!(lines[1].starts_with("build1"));
+            }
+            assert!(lines[2].starts_with("dependencies"));
+        },
+    );
+
+    rust_child_test_with_assertion_fn(
+        "list_dependencies_with_alternate_registry_restriction_in_workspace_with_unset_alternate_registry",
+        TestVariants::AlternateRegistryRestrictionInWorkspaceUnsetAlt,
+        |output| {
+            let stderr = String::from_utf8_lossy(&output.stderr).to_string();
+            let lines: Vec<_> = stderr.lines().collect();
+
+           assert!(lines.is_empty()) 
+        },
+    );
+
+    rust_child_test_with_assertion_fn(
+        "list_dependencies_with_alternate_registry_restriction_in_workspace",
+        TestVariants::AlternateRegistryRestrictionInWorkspace,
+        |output| {
+            let stderr = String::from_utf8_lossy(&output.stderr).to_string();
+            let lines: Vec<_> = stderr.lines().collect();
+
+            if lines[0].starts_with("build1") {
+                assert!(lines[1].starts_with("dep1"), "{}", &lines.join("\n"));
+            } else {
+                assert!(lines[0].starts_with("dep1"));
+                assert!(lines[1].starts_with("build1"));
+            }
+            assert!(lines[2].starts_with("dependencies_alt_registry"));
+        },
+    );
+
+    println!(
+        "\ntest result: ok. {} passed; 0 failed; 0 ignored; 0 measured; 0 filtered out",
+        test_cnt
+    )
+}
```

Comment:
This is worth calling out as I spent a lot of time trying to think on the best way to approach testing the logging that was compatible with the current tests. Best I could come up with was something I found on the [env_logger regexp_filter tests](https://github.com/rust-cli/env_logger/blob/main/tests/regexp_filter.rs) that made a custom harness that executed itself to make the logger unique per invocation.

This works but feels hacky, I'm very open to alternate ideas.

Reply:
It's a neat approach that I hadn't considered using for this sort of thing. Off the top of my head, I can't beat it! I'd rather have hacky tests than no tests, so thank you for spending some time investigating ways forward 🙂